content/post/2019-01-18-nerves-at-home-desk-controller.md

```diff
@@ -0,0 +1,245 @@
+incercept current height messages from the controller then publish to a phoenix
+site for real-time desk height measurements...because #yolo.
+
+<!--more-->
+
+## The idea
+
+I'm fairly obsessed with automating my home and things around me. With nerves,
+that has become even easier and much more cost effective. But most of my recent 
+work has been pretty binary - turning relay on/off, reading a sensor is on/off,
+reading the temperature, etc etc. I wanted to branch out a bit and try to
+directly interface with an existing component in my house. Something that is it's
+own controller and has its own message system.
+
+One day I leaned back to take a break from work, looked over at my [Uplift Desk
+Advanced Digital Memory Keypad](https://www.upliftdesk.com/advanced-digital-memory-keypad/)
+for my sit/stand desk and it hit me - _"I bet I could control that with nerves..."_
+🤔
+
+And thus this project was born (albeit useless as it is...)
+
+## Interfacing with the desk
+
+Mine is a 2-leg model from Uplift Desk which has their standard [Smart Control Box]
+(https://www.upliftdesk.com/desk-control-box-for-2-leg-3-leg-or-4-leg-desks/).
+The description states it _"Houses the 'brain' of your UPLIFT Desk"_ along with
+other settings like motor speed, min/max heights, and saved height memory. Which
+is great because that means the [Advanced Digital Memory Keypad](https://www.upliftdesk.com/advanced-digital-memory-keypad/)
+is essentially just a simple microcontroller receiving and sending messages with
+"the brain", so we can intercept there with the nerves controller.
+
+The keypad connects to the desk controller via a RJ45 jack and looks like a
+typical ethernet cable with an 8-wire pinout. So I took apart to see what the
+connector looked like for those 8 wires and was pleasently surprised. The wires
+were clearly colored and the PCB had a label for each one so I was able to immediately
+identify `5V`, `GND`, and `G1`. The other wire labels were mostly covered by the
+connector, but I was able to make out `T`, `R`, and `G0` as well.
+
+![desk_keypad_internals](/images/2019-01-18/desk_keypad_internals.jpg)
+
+So I set out with a multimeter to map out the wires as best I could. `T`/`R` were
+most likely short for `Tx`/`Rx` and would be the communication with the controller,
+so I skipped those for now. Here's the final pinout I came up with:
+
+![desk_keypad_pinout](/images/2019-01-18/desk_keypad_pinout.png)
+
+Moving the desk is simply just connecting the yellow or green wire to GND. So that
+can be easily done by toggling a GPIO pin between _HIGH (1)_ and _LOW (0)_ using
+[Circuits.GPIO](https://github.com/elixir-circuits/circuits_gpio).
+
+```elixir
+# setting LOW will move the desk in whatever direction the pin is for
+# Be sure to initialize with HIGH to prevent moving on pin open.
+{:ok, up_pin} = Circuits.GPIO.open(23, :output, initial_value: 1)
+{:ok, down_pin} = Circuits.GPIO.open(24, :output, initial_value: 1)
```

Comment:
Yay! You're using the new `initial_value` option.

Reply:
Ya, I was the one that wrote it into ElixirALE so was really glad you guys wrote it in new circuits.GPIO...really saved me!